service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,253 @@
+                                  const size_t nColumns, const size_t nThreads, const size_t blockSize)
+{
+    services::Status s;
+
+    if (nRows * nColumns > THREADING_BORDER && nThreads > 1)
+    {
+        const size_t nBlocks = nRows / blockSize + !!(nRows % blockSize);
+
+        daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
+            const size_t start = iBlock * blockSize;
+            const size_t end   = daal::services::internal::min<cpu, size_t>(start + blockSize, nRows);
+
+            s |= assignRows<DataType, IdxType>(origDataPtr, dataTable, idxTable, start, end - start, nColumns);
+        });
+    }
+    else
+    {
+        s |= assignRows<DataType, IdxType>(origDataPtr, dataTable, idxTable, 0, nRows, nColumns);
+    }
+
+    return s;
+}
+
+template <typename DataType, typename IdxType, daal::CpuType cpu>
+services::Status splitRows(const NumericTablePtr inputTable, const NumericTablePtr trainTable, const NumericTablePtr testTable,
+                           const NumericTablePtr trainIdxTable, const NumericTablePtr testIdxTable, const size_t nTrainRows, const size_t nTestRows,
+                           const size_t nColumns, const size_t nThreads)
+{
+    services::Status s;
+    const size_t blockSize = daal::services::internal::max<cpu, size_t>(BLOCK_CONST / nColumns, 1);
+    BlockDescriptor<DataType> origBlock;
+    inputTable->getBlockOfRows(0, nTrainRows + nTestRows, readOnly, origBlock);
+    const DataType * origDataPtr = origBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(origDataPtr);
+
+    s |= assignRowsSubset<DataType, IdxType, cpu>(origDataPtr, trainTable, trainIdxTable, nTrainRows, nColumns, nThreads, blockSize);
+    s |= assignRowsSubset<DataType, IdxType, cpu>(origDataPtr, testTable, testIdxTable, nTestRows, nColumns, nThreads, blockSize);
+
+    return s;
+}
+
+template <typename IdxType, daal::CpuType cpu>
+services::Status trainTestSplitImpl(const NumericTablePtr inputTable, const NumericTablePtr trainTable, const NumericTablePtr testTable,
+                                    const NumericTablePtr trainIdxTable, const NumericTablePtr testIdxTable)
+{
+    services::Status s;
+    const size_t nThreads   = threader_get_threads_number();
+    const NTLayout layout   = inputTable->getDataLayout();
+    const size_t nColumns   = trainTable->getNumberOfColumns();
+    const size_t nTrainRows = trainTable->getNumberOfRows();
+    const size_t nTestRows  = testTable->getNumberOfRows();
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, nTrainRows + nTestRows, nColumns);
+
+    NumericTableDictionaryPtr tableFeaturesDict = inputTable->getDictionarySharedPtr();
+
+    if (layout == NTLayout::soa)
+    {
+        BlockDescriptor<IdxType> trainIdxBlock, testIdxBlock;
+        trainIdxTable->getBlockOfColumnValues(0, 0, nTrainRows, readOnly, trainIdxBlock);
```

Comment:
Are you sure that there are not potential conversion in seq code?

Reply:
Conversion will be there, but it's hotspot only for extreme dimensions (many rows, few columns).

Reply:
What is a problem to move this under parallel region?

Reply:
Difficulties to support it for sequential and threading branches, additional code complications

Reply:
Okay, I just want to avoid the same situation as with data conversion in finite checks.